File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/SemanticModelFingerprint.java
package com.foundgine.core.semantic;

import com.foundgine.core.abstractions.*;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.util.*;

/** Canonical provider-independent SHA-256 fingerprint of a semantic model. */
public final class SemanticModelFingerprint {
    private SemanticModelFingerprint() {}
    public static String compute(SemanticModel model) {
        Objects.requireNonNull(model);
        var canonical=new StringBuilder("foundgine.semantic-contract.v1\n");
        model.entities().stream().sorted(Comparator.comparingLong(e -> e.id().value())).forEach(e -> {
            append(canonical,"entity",e.id().value()); append(canonical,"name",e.name());
            append(canonical,"identity",e.identity().fieldId().value(),e.identity().name()); appendAliases(canonical,e.effectiveAliases());
            e.fields().stream().sorted(Comparator.comparingLong(f -> f.id().value())).forEach(f -> {
                append(canonical,"field",f.id().value()); append(canonical,"field-name",f.name()); append(canonical,"field-type",canonicalType(f.effectiveSemanticType()));
                append(canonical,"field-capabilities",f.capabilities()); append(canonical,"field-nullable",f.isNullable()); appendAliases(canonical,f.effectiveAliases());
                f.effectiveConstraints().stream().sorted(Comparator.comparing((SemanticConstraint c)->c.kind().ordinal()).thenComparing(c->Objects.toString(c.value(),"")).thenComparing(c->Objects.toString(c.minimum(),"")).thenComparing(c->Objects.toString(c.maximum(),""))).forEach(c -> append(canonical,"constraint",c.kind().ordinal(),Objects.toString(c.value(),""),decimal(c.minimum()),decimal(c.maximum())));
            });
            e.relationships().stream().sorted(Comparator.comparingLong(r -> r.id().value())).forEach(r -> { append(canonical,"relationship",r.id().value(),r.name(),r.target().value(),r.cardinality().ordinal()); appendAliases(canonical,r.effectiveAliases()); });
        });
        model.traversals().stream().sorted(Comparator.comparingLong((SemanticTraversal t)->t.source().value()).thenComparing(SemanticTraversal::name)).forEach(t -> { append(canonical,"traversal",t.source().value(),t.name(),t.target().value()); t.path().forEach(r -> append(canonical,"traversal-edge",r.value())); });
        try { var d=MessageDigest.getInstance("SHA-256"); var bytes=d.digest(canonical.toString().getBytes(StandardCharsets.UTF_8)); var out=new StringBuilder(64); for(byte b:bytes) out.append(String.format("%02x",b)); return out.toString(); } catch(Exception ex){ throw new IllegalStateException(ex); }
    }
    private static String decimal(java.math.BigDecimal v){ return v==null?"":v.stripTrailingZeros().toPlainString(); }
    private static void appendAliases(StringBuilder b,List<SemanticAlias> a){ a.stream().sorted(Comparator.comparing(SemanticAlias::name,String.CASE_INSENSITIVE_ORDER).thenComparing(SemanticAlias::name)).forEach(x->append(b,"alias",x.name(),x.weight()==null?"":x.weight())); }
    private static void append(StringBuilder b,String kind,Object... values){ b.append(kind); for(Object v:values){String s=String.valueOf(v); b.append('|').append(s.length()).append(':').append(s);} b.append('\n'); }
    private static String canonicalType(SemanticType t){ return switch(t){ case SemanticType.Scalar x -> "scalar:"+x.kind(); case SemanticType.EnumType x -> "enum:"+x.name(); case SemanticType.ObjectType x -> "object:"+x.name(); case SemanticType.CollectionType x -> "collection:"+canonicalType(x.elementType()); }; }
}

File: src/csharp/Foundgine.Core/Semantic/SemanticModelFingerprint.cs
using System.Security.Cryptography;
using System.Text;
using Foundgine.Core.Abstractions;

namespace Foundgine.Core.Semantic;

/// <summary>
/// Produces a canonical, provider-independent fingerprint for a semantic model.
/// The fingerprint is based only on semantic declarations and is independent of
/// registration/declaration order or CLR metadata identity.
/// </summary>
public static class SemanticModelFingerprint
{
    public static string Compute(SemanticModel model)
    {
        ArgumentNullException.ThrowIfNull(model);

        var canonical = new StringBuilder("foundgine.semantic-contract.v1\n");

        foreach (var entity in model.Entities.OrderBy(x => x.Id.Value))
        {
            Append(canonical, "entity", entity.Id.Value);
            Append(canonical, "name", entity.Name);
            Append(canonical, "identity", entity.Identity.FieldId.Value, entity.Identity.Name);
            AppendAliases(canonical, entity.EffectiveAliases);

            foreach (var field in entity.Fields.OrderBy(x => x.Id.Value))
            {
                Append(canonical, "field", field.Id.Value);
                Append(canonical, "field-name", field.Name);
                Append(canonical, "field-type", CanonicalType(field.EffectiveSemanticType));
                Append(canonical, "field-capabilities", (byte)field.Capabilities);
                Append(canonical, "field-nullable", field.IsNullable);
                AppendAliases(canonical, field.EffectiveAliases);

                foreach (var constraint in field.EffectiveConstraints
                             .OrderBy(x => x.Kind)
                             .ThenBy(x => x.Value, StringComparer.Ordinal)
                             .ThenBy(x => x.Minimum)
                             .ThenBy(x => x.Maximum))
                {
                    Append(canonical, "constraint", (byte)constraint.Kind, constraint.Value ?? "",
                        constraint.Minimum?.ToString("G29", System.Globalization.CultureInfo.InvariantCulture) ?? "",
                        constraint.Maximum?.ToString("G29", System.Globalization.CultureInfo.InvariantCulture) ?? "");
                }
            }

            foreach (var relationship in entity.Relationships.OrderBy(x => x.Id.Value))
            {
                Append(canonical, "relationship", relationship.Id.Value, relationship.Name,
                    relationship.Target.Value, (byte)relationship.Cardinality);
                AppendAliases(canonical, relationship.EffectiveAliases);
            }
        }

        foreach (var traversal in model.Traversals
                     .OrderBy(x => x.Source.Value)
                     .ThenBy(x => x.Name, StringComparer.Ordinal))
        {
            Append(canonical, "traversal", traversal.Source.Value, traversal.Name, traversal.Target.Value);
            foreach (var relationshipId in traversal.Path)
                Append(canonical, "traversal-edge", relationshipId.Value);
        }

        var bytes = SHA256.HashData(Encoding.UTF8.GetBytes(canonical.ToString()));
        return Convert.ToHexString(bytes).ToLowerInvariant();
    }

    private static void AppendAliases(StringBuilder builder, IReadOnlyList<SemanticAlias> aliases)
    {
        foreach (var alias in aliases.OrderBy(x => x.Name, StringComparer.OrdinalIgnoreCase)
                     .ThenBy(x => x.Name, StringComparer.Ordinal))
            Append(builder, "alias", alias.Name,
                alias.Weight?.ToString(System.Globalization.CultureInfo.InvariantCulture) ?? "");
    }

    private static void Append(StringBuilder builder, string kind, params object[] values)
    {
        builder.Append(kind);
        foreach (var value in values)
        {
            var text = value.ToString() ?? "";
            builder.Append('|').Append(text.Length).Append(':').Append(text);
        }

        builder.Append('\n');
    }

    private static string CanonicalType(SemanticType type) => type switch
    {
        SemanticType.Scalar scalar => $"scalar:{scalar.Kind}",
        SemanticType.Enum value => $"enum:{value.Name}",
        SemanticType.Object value => $"object:{value.Name}",
        SemanticType.Collection value => $"collection:{CanonicalType(value.ElementType)}",
        _ => throw new InvalidOperationException($"Unsupported semantic type '{type.GetType().FullName}'.")
    };
}